### ISSUE TITLE
Undefined reference to '_M_get_insert_unique_pos' (FixedLenDecoderEmitter.cpp.o.elf) with existing clang.db

### ISSUE THREAD

**Issue by MaggieYingYi** [SNSystems/llvm-project-prepo#19]
This is spin-off from the issue SNSystems/llvm-project-prepo#7.

Following the build steps mentioned in <https://github.com/SNSystems/llvm-project-prepo/issues/7#issuecomment-478892314>, the repo toolchain passed the LLVM first time build.

Clean and rebuild the LLVM using the repo toolchain with existing clang.db.
```
$ cd llvm-project-prepo
$ ninja clean
$ ninja
```
Failed with the following linking error.
```
ERROR:wrap_tool:b"utils/TableGen/CMakeFiles/obj.llvm-tblgen.dir/FixedLenDecoderEmitter.cpp.o.elf: In function `std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<(anonymous namespace)::OperandInfo, std::allocator<(anonymous namespace)::OperandInfo> > >, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<(anonymous namespace)::OperandInfo, std::allocator<(anonymous namespace)::OperandInfo> > > >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<(anonymous namespace)::OperandInfo, std::allocator<(anonymous namespace)::OperandInfo> > > > >::_M_get_insert_hint_unique_pos(std::_Rb_tree_const_iterator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<(anonymous namespace)::OperandInfo, std::allocator<(anonymous namespace)::OperandInfo> > > >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':\n(.text+0x15de0): undefined reference to `std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<(anonymous namespace)::OperandInfo, std::allocator<(anonymous namespace)::OperandInfo> > >, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<(anonymous namespace)::OperandInfo, std::allocator<(anonymous namespace)::OperandInfo> > > >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<(anonymous namespace)::OperandInfo, std::allocator<(anonymous namespace)::OperandInfo> > > > >::_M_get_insert_unique_pos(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
```

**Comment by MaggieYingYi** [comment]
Analysis of the issue: build the ‘FixedLenDecoderEmitter.cpp’ (first time).

Step 1: Compiled ‘FixedLenDecoderEmitter.cpp’ with a command something like:
```
$ /home/user/github/llvm-project-prepo/build/bin/clang++  --target=x86_64-pc-linux-gnu-repo  -DGTEST_HAS_RTTI=0 -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Iutils/TableGen -I/home/user/github/llvm-project-prepo/llvm/utils/TableGen -Iinclude -I/home/user/github/llvm-project-prepo/llvm/include -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -std=c++11 -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wstring-conversion -fdiagnostics-color -ffunction-sections -fdata-sections  -O3 -fno-exceptions -fno-rtti -MMD -MT utils/TableGen/CMakeFiles/obj.llvm-tblgen.dir/FixedLenDecoderEmitter.cpp.o -MF utils/TableGen/CMakeFiles/obj.llvm-tblgen.dir/FixedLenDecoderEmitter.cpp.o.d -o utils/TableGen/CMakeFiles/obj.llvm-tblgen.dir/FixedLenDecoderEmitter.cpp.1.o -c /home/user/github/llvm-project-prepo/llvm/utils/TableGen/FixedLenDecoderEmitter.cpp
```

Step 2: Get the compilation hash in object file (FixedLenDecoderEmitter.cpp.1.o).
```
$ /home/user/github/llvm-project-prepo/build/bin/repo-ticket-dump utils/TableGen/CMakeFiles/obj.llvm-tblgen.dir/FixedLenDecoderEmitter.cpp.1.o
635e304a80043bcd29b30bc1b25f8827
```

Step 3: Find the fragments named '_ZNSt8_Rb_treeINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_St6vectorIN12_GLOBAL__N_111OperandInfoESaISA_EEESt10_Select1stISD_ESt4lessIS5_ESaISD_EE29_M_get_insert_hint_unique_posESt23_Rb_tree_const_iteratorISD_ERS7_' and '_ZNSt8_Rb_treeINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_St6vectorIN12_GLOBAL__N_111OperandInfoESaISA_EEESt10_Select1stISD_ESt4lessIS5_ESaISD_EE24_M_get_insert_unique_posERS7_' in the compilation.

```
$ /home/user/github/llvm-project-prepo/build/bin/pstore-dump -compilation=635e304a80043bcd29b30bc1b25f8827 clang.db
---
- file         : 
      path : clang.db
      size : 4194304
  compilations : 
      - digest      : 635e304a80043bcd29b30bc1b25f8827
        compilation : 
            members : 
               :
			   :
               - digest  : d05a1bdf24bf3f51ea9c67365705b469
                 fext    : { addr: 401344, size: 888 }
                 name    : _ZNSt8_Rb_treeINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_St6vectorIN12_GLOBAL__N_111OperandInfoESaISA_EEESt10_Select1stISD_ESt4lessIS5_ESaISD_EE29_M_get_insert_hint_unique_posESt23_Rb_tree_const_iteratorISD_ERS7_
                 linkage : internal             
               - digest  : 1f42cfc7d095a402568a5d494855c2e4
                 fext    : { addr: 470448, size: 424 }
                 name    : _ZNSt8_Rb_treeINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_St6vectorIN12_GLOBAL__N_111OperandInfoESaISA_EEESt10_Select1stISD_ESt4lessIS5_ESaISD_EE24_M_get_insert_unique_posERS7_
                 linkage : internal
				 
```

Step 4: Dump the fragment named '_ZNSt8_Rb_treeINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_St6vectorIN12_GLOBAL__N_111OperandInfoESaISA_EEESt10_Select1stISD_ESt4lessIS5_ESaISD_EE29_M_get_insert_hint_unique_posESt23_Rb_tree_const_iteratorISD_ERS7_' in the database.
```
$ /home/user/github/llvm-project-prepo/build/bin/pstore-dump -fragment=d05a1bdf24bf3f51ea9c67365705b469 clang.db
---
- file      : 
      path : clang.db
      size : 4194304
  fragments : 
      :
	  :
      - digest   : d05a1bdf24bf3f51ea9c67365705b469
        fragment : 
            - type     : text
              contents : 
                  align   : 16
                  data    : 
                            …
                  ifixups : [ ]
                  xfixups : 
                      …
                      - name   : _ZNSt8_Rb_treeINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_St6vectorIN12_GLOBAL__N_111OperandInfoESaISA_EEESt10_Select1stISD_ESt4lessIS5_ESaISD_EE24_M_get_insert_unique_posERS7_
                        type   : 4
                        offset : 512
                        addend : 18446744073709551612
            - type     : dependent
              contents : 
                  - digest  : 1f42cfc7d095a402568a5d494855c2e4
                    fext    : { addr: 470448, size: 424 }
                    name    : _ZNSt8_Rb_treeINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_St6vectorIN12_GLOBAL__N_111OperandInfoESaISA_EEESt10_Select1stISD_ESt4lessIS5_ESaISD_EE24_M_get_insert_unique_posERS7_
                    linkage : internal

```


In summary, the fragment of '_ZNSt8_Rb_treeINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_St6vectorIN12_GLOBAL__N_111OperandInfoESaISA_EEESt10_Select1stISD_ESt4lessIS5_ESaISD_EE29_M_get_insert_hint_unique_posESt23_Rb_tree_const_iteratorISD_ERS7_' is dependent on the fragment of '_ZNSt8_Rb_treeINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_St6vectorIN12_GLOBAL__N_111OperandInfoESaISA_EEESt10_Select1stISD_ESt4lessIS5_ESaISD_EE24_M_get_insert_unique_posERS7_'.

**Comment by MaggieYingYi** [comment]
Rebuild the ‘FixedLenDecoderEmitter.cpp’ again (second time).

Step 1: Compiled ‘FixedLenDecoderEmitter.cpp’ with the same command mentioned in https://github.com/SNSystems/llvm-project-prepo/issues/19#issuecomment-479532503, but output to 'FixedLenDecoderEmitter.cpp.2.o'.

Step 2: Get the compilation_member in the object file (FixedLenDecoderEmitter.cpp.2.o).
```
$ /home/user/github/llvm-project-prepo/build/bin/repo-ticket-dump utils/TableGen/CMakeFiles/obj.llvm-tblgen.dir/FixedLenDecoderEmitter.cpp.2.o
bd659a27e3ceda9f9a13d8d576f9fe0a
$ /home/user/github/llvm-project-prepo/build/bin/pstore-dump -compilation=bd659a27e3ceda9f9a13d8d576f9fe0a clang.db
---
- file         : 
      path : clang.db
      size : 4194304
  compilations : 
      - digest      : 635e304a80043bcd29b30bc1b25f8827
        compilation : 
            members : 
               :
			   :
                - digest  : d05a1bdf24bf3f51ea9c67365705b469
                  fext    : { addr: 401344, size: 888 }
                  name    : _ZNSt8_Rb_treeINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_St6vectorIN12_GLOBAL__N_111OperandInfoESaISA_EEESt10_Select1stISD_ESt4lessIS5_ESaISD_EE29_M_get_insert_hint_unique_posESt23_Rb_tree_const_iteratorISD_ERS7_
                  linkage : internal
                - digest  : d05a1bdf24bf3f51ea9c67365705b469
                  fext    : { addr: 401344, size: 888 }
                  name    : _ZNSt8_Rb_treeINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_St6vectorIN12_GLOBAL__N_111OperandInfoESaISA_EEESt10_Select1stISD_ESt4lessIS5_ESaISD_EE29_M_get_insert_hint_unique_posESt23_Rb_tree_const_iteratorISD_ERS7_
                  linkage : internal

				 
```


There are two issues in the compilation members for the second time build.
```
1) The fragment of _ZNSt8_Rb_treeINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_St6vectorIN12_GLOBAL__N_111OperandInfoESaISA_EEESt10_Select1stISD_ESt4lessIS5_ESaISD_EE29_M_get_insert_hint_unique_posESt23_Rb_tree_const_iteratorISD_ERS7_ was put twice into the compilation members.
2) The fragment of _ZNSt8_Rb_treeINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_St6vectorIN12_GLOBAL__N_111OperandInfoESaISA_EEESt10_Select1stISD_ESt4lessIS5_ESaISD_EE24_M_get_insert_unique_posERS7_ was missed in the compilation members during second build. This caused the above error.
```

**Comment by MaggieYingYi** [comment]
To reproduce the issue, compile the following IR code (test.ll):
```
target triple = "x86_64-pc-linux-gnu-repo"

define internal i32 @A()  !repo_ticket !0 {
entry:
  %call = call i32 @B()
  ret i32 %call
}

define internal i32 @B() {
entry:
  %call = call i32 @C()
  ret i32 %call
}

define internal i32 @C() {
entry:
  ret i32 1
}

!repo.tickets = !{!0}
!0 = !TicketNode(name: "A", digest: [16 x i8] c"m\EB7\F7\055\B2\DD\D5\B3\FAX<o\E5\87", linkage: internal, pruned: false)
```

Compiled to the repo with a command something like:
```
$ llc -filetype=obj test.ll -o test.o
```

Produces compilations records:
```
$ pstore-dump -all-compilations clang.db
---

- file         :
      path : clang.db
      size : 4194304
  compilations :
      - digest      : 7c2ba702542dba4c2c80932eb45b7829
        compilation :
            members :
                - digest  : 87e56f3c58fab3d5ddb23505f737eb6d
                  fext    : { addr: 400, size: 96 }
                  name    : A
                  linkage : internal
                - digest  : 7e18bff7d2be8e1fed1d7897ead7b120
                  fext    : { addr: 272, size: 96 }
                  name    : B
                  linkage : internal
                - digest  : b58747adb3c5dceb13cabf0d0df183f5
                  fext    : { addr: 528, size: 38 }
                  name    : C
                  linkage : internal
            path    : \\\\?\\C:\\MyWork\\repo_bug\\issue19\\test
            triple  : x86_64-pc-linux-gnu-repo

```
Re-build the test.ll again, there shouldn’t be any database transaction happen since the test.ll doesn’t change. However, a new file ticket has been insert to the database.
```
$ clang -O3 -c --target=x86_64-pc-linux-gnu-repo -x ir test.ll  -o test.o
$ pstore-dump -all-compilations clang.db
---

- file         :
      path : clang.db
      size : 4194304
  compilations :
     ...
      - digest      : 2428b53078c95dff71dab06cbbf857b5
        compilation :
            members :
                - digest  : 87e56f3c58fab3d5ddb23505f737eb6d
                  fext    : { addr: 400, size: 96 }
                  name    : A
                  linkage : internal
                - digest  : 87e56f3c58fab3d5ddb23505f737eb6d
                  fext    : { addr: 400, size: 96 }
                  name    : A
                  linkage : internal
                - digest  : 7e18bff7d2be8e1fed1d7897ead7b120
                  fext    : { addr: 272, size: 96 }
                  name    : B
                  linkage : internal
            path    : \\\\?\\C:\\MyWork\\repo_bug\\issue19\\test
            triple  : x86_64-pc-linux-gnu-repo
```
The new compilation member (2428b53078c95dff71dab06cbbf857b5) has two issues:
```
1) The function A was put twice into the compilation members.
2) The function C was missed in the compilation members.
```

**Comment by MaggieYingYi** [comment]
Passed code review (#20).

Fixed in the git commit: 2b6afb4.
